backend/main/shared/dps.go

```diff
@@ -0,0 +1,213 @@
+	"encoding/json"
+	"io/ioutil"
+	"strings"
+
+	"github.com/onflow/cadence"
+	dpsApi "github.com/onflow/flow-dps/api/dps"
+	"github.com/onflow/flow-dps/codec/zbor"
+	"github.com/onflow/flow-dps/service/invoker"
+	"github.com/onflow/flow-go-sdk"
+	_flow "github.com/onflow/flow-go/model/flow"
+	"github.com/rs/zerolog/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+var (
+	DummyBalance = FTBalanceResponse{
+		PrimaryAccountBalance:   100,
+		SecondaryAccountBalance: 100,
+		StakingBalance:          100,
+		BlockHeight:             0,
+	}
+)
+
+type DpsAdapter struct {
+	Config  FlowConfig
+	Invoker *invoker.Invoker
+	Context context.Context
+	URL     string
+	Env     string
+}
+
+func NewDpsClient(flowEnv string) *DpsAdapter {
+	// Initialize adapter
+	adapter := DpsAdapter{}
+	adapter.Context = context.Background()
+	adapter.Env = flowEnv
+
+	// parse flow.json
+	path := "./flow.json"
+
+	content, err := ioutil.ReadFile(path)
+
+	if err != nil {
+		log.Fatal().Msgf("Error when opening file: %+v.", err)
+	}
+
+	var config FlowConfig
+	err = json.Unmarshal(content, &config)
+	if err != nil {
+		log.Fatal().Msgf("Error parsing flow.json: %+v.", err)
+	}
+
+	adapter.Config = config
+	adapter.URL = config.DpsNetworks[adapter.Env]
+
+	// For emulator, skip invoker instantiation
+	if adapter.URL == "" {
+		return &adapter
+	}
+
+	// Initialize the DPS API client.
+	codec := zbor.NewCodec()
+
+	conn, err := grpc.DialContext(
+		adapter.Context,
+		adapter.URL,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.FailOnNonTempDialError(true), // DialContext will fail immediately if passed bad connection string
+		grpc.WithBlock(),                  // connection should be blocking
+	)
+
+	if err != nil {
+		log.Error().Str("dps", adapter.URL).Err(err).Msg("could not dial API host")
+		panic("error initializing DPS gRPC connection")
+	}
+
+	client := dpsApi.NewAPIClient(conn)
+	index := dpsApi.IndexFromAPI(client, codec)
+
+	cacheSize := uint64(100_000_000)
+	_invoker, err := invoker.New(index, invoker.WithCacheSize(cacheSize))
+	adapter.Invoker = _invoker
+
+	if err != nil {
+		log.Error().Str("dps", adapter.URL).Err(err)
+		panic("error initializing DPS Invoker")
+	}
+	return &adapter
+}
+
+// FLOW token has its own function for fetching balance because
+// it has its own script to account for locked/staked tokens.
+func (dps *DpsAdapter) GetFlowBalanceAtBlockheightScript(addr string, blockheight uint64) (cadence.Value, error) {
+	var args []cadence.Value
+
+	// Prepare Args
+	flowAddress := flow.HexToAddress(addr)
+	cadenceAddress := cadence.NewAddress(flowAddress)
+
+	args = append(args, cadenceAddress)
+
+	_script, err := ioutil.ReadFile("./main/cadence/scripts/get_flow_token_balances.cdc")
+
+	if err != nil {
+		log.Error().Err(err).Msgf("Error reading cadence script file.")
+		return nil, err
+	}
+
+	// Replace script args
+	script := dps.Config.InsertCoreContractAddresses(string(_script))
+
```

Comment:
the clean up here seems good with `InsertCoreContractAddresses`!

Reply:
It would be good if `InsertCoreContractAddresses` could return an error, incase one of those values is removed from the flow.json. Could be a nasty thing to debug. ..
`ReplaceAllString` doesn't natively return an error, but maybe there is a way to error handle this logic?

Reply:
Yeah I'm not sure... I was definitely debugging it when I was writing it.  If a placeholder doesnt get replaced, or is replaced incorrectly, I was getting a script contract error from the emulator.  Its not the _worst_ thing to debug.  I'm just not sure what the right way to detect an error is in this case.

Reply:
ah ok in that case, if we're getting error returned I think its ok.